Allow running just load calculations and no full simulations

EnergyArchmage opened the pull request:
Pull request overview
---------------------
 - Fixes #9191
 - Interface developer requests ability to just run load calculations and no full simulations of design days or run periods.  Once the ideal loads sizing calculations are completed, and the simulation setup has occurred, the component loads and component sizes are known.  Users may not care about energy and operation performance and just want loads and equipment sizes.  

Previously models would not run if SimulationControl fields for Run Simulation for Sizing Periods and Run Simulation for Weather File Run Periods were both set to No (despite what the I/O reference seems to say; issue #9191). With this pull request, these inputs can both be No and the model will just run the ideal loads sizing calculations for the design days and then stop and report what it can. 

This has important run time performance advantage to shorten execution. Since full simulations of design days have HVAC, the Ideal loads calculations run faster than full simulations of design days.  Performance improvement is on the order of 3X. 

follow on from #9386 , @Myoldmopar , @jmarrec 

Note that, this also turns on detailed time series reporting (from Output:Variable) during zone sizing calculations when doing a pure load calculation.  This provides a wealth of information on zone heat balance calculations during ideal loads sizing that can provide context to explain component loads reporting. 

Contribution from TRANE

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

EnergyArchmage committed:
enable runs with just load calculations and no full simulation of environments

EnergyArchmage committed:
fix clang format in SimulationManager.cc

Myoldmopar commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -5282,9 +5282,12 @@ void WriteTabularReports(EnergyPlusData &state)
+        if (ort->displayCO2ResilienceSummary && !state.dataGlobal->DoPureLoadCalc)     // code will crash if pure load calc
+            WriteCO2ResilienceTables(state);                                           // code will crash if pure load calc
+        if (ort->displayVisualResilienceSummary && !state.dataGlobal->DoPureLoadCalc)  // code will crash if pure load calc
+            WriteVisualResilienceTables(state);                                        // code will crash if pure load calc
```
Is it worth issuing a warning for these cases when the user has possibly requested the outputs but we aren't generating them for the given run configuration?

EnergyArchmage replied:
I don't think so.  There already will be warnings for Output:Variable requests that are present but cannot be fulfilled because the program never got that far.   

I have now added detailed timeseries data reporting for the ideal loads sizing calculations.  The zone heat balance data are very useful for developing a fuller understanding of the component loads.

Myoldmopar replied:
:+1:

Myoldmopar commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -1153,6 +1164,12 @@ namespace SimulationManager {
+             !state.dataGlobal->DoHVACSizingSimulation) &&                          // pure load calcs option
+            (state.dataGlobal->DoZoneSizing || state.dataGlobal->DoSystemSizing)) { // pure load calcs option
+            state.dataGlobal->DoPureLoadCalc = true;                                // pure load calcs option
+        }                                                                           // pure load calcs option
```
Although it's not the typical way we use references to object instances, this could be quite readable as:

```
    auto &dg = state.dataGlobal;
    if ((!dg->DoDesDaySim && !dg->DoWeathSim && !dg->DoHVACSizingSimulation) && (dg->DoZoneSizing || dg->DoSystemSizing)) {
        dg->DoPureLoadCalc = true;
    }
```

I would personally say that is way more readable, but if anyone feels that is not a great use of a reference, then no big deal.

Myoldmopar commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -304,6 +304,12 @@ namespace SimulationManager {
+            state.dataGlobal->DoOutputReporting = true;    // pure load calcs option
+            Available = false;                             // pure load calcs option, don't run any environments for final simulation
+            state.dataOutRptTab->WriteTabularFiles = true; // pure load calcs option, need tabular reports for component loads tables
+        }                                                  // pure load calcs option
```
I think prior to merge we'll clean out the repeated comment block.

Myoldmopar reviewed (commented):
A few minor comments but this will certainly be a nice addition.  I'm happy to do some final polishing on this, but would like at least a reply about the warning before proceeding.  And this would need to go in ahead of anything from #9386, right?

EnergyArchmage committed:
add reporting of timeseries data during ideal loads sizing if only doing load calcs

enable Output:Variable reporting during sizing if doing a pure load calc run.
enable detailed warmup convergence reporting if doing a pure load calc run.
enable ReportDuringWarmup  if doing a pure load calc run.

EnergyArchmage committed:
Merge branch 'develop' into 9191-pure-loads-calc

jmarrec commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -1123,6 +1131,11 @@ namespace SimulationManager {
             state.dataGlobal->DoWeathSim = true;
         }
 
+        if ((!state.dataGlobal->DoDesDaySim && !state.dataGlobal->DoWeathSim && !state.dataGlobal->DoHVACSizingSimulation) &&
+            (state.dataGlobal->DoZoneSizing || state.dataGlobal->DoSystemSizing)) {
+            state.dataGlobal->DoPureLoadCalc = true;
+        }
```
Do we need to add a condition for DoPlantSizing here?
```suggestion
        if ((!state.dataGlobal->DoDesDaySim && !state.dataGlobal->DoWeathSim && !state.dataGlobal->DoHVACSizingSimulation) &&
            (state.dataGlobal->DoZoneSizing || state.dataGlobal->DoSystemSizing || state.dataGlobal->DoPlantSizing)) {
            state.dataGlobal->DoPureLoadCalc = true;
        }
```

EnergyArchmage replied:
I don't believe it is needed.   The use case here doesn't really include doing only plant sizing.

Myoldmopar commented:
Looks like there are a couple unit tests that fail.  Presumably those unit tests need some flag to be set or something.  I'll try to reproduce locally first.

EnergyArchmage commented:
Should probably include some new or changed integration test files so this gets included in CI

Myoldmopar commented:
It looks to be aborting at https://github.com/NREL/EnergyPlus/blob/b300d66751982a6f8e220e16eec31cd6e5a4de88/src/EnergyPlus/OutputProcessor.cc#L5630

It's trying to access the TimeValue map using the `TimeStepType::System` key, but that map isn't initialized with that.  There are plenty of places in the unit test code that [initialize the map](https://github.com/NREL/EnergyPlus/blob/b300d66751982a6f8e220e16eec31cd6e5a4de88/tst/EnergyPlus/unit/OutputProcessor.unit.cc#L5204-L5205), so I guess maybe we just need to add that to this failing unit test?  Or does this indicate that this new bypass route is actually missing something important.

I agree we should add an IDF integration test that covers this.

rraustad commented:
Here's how I fixed a unit test with the same problem.

I added this to the unit test right before the call to ManageSizing (or whatever unit test function call gets to the problem) and this unit test now runs.

state->dataLoopNodes->MoreNodeInfo.allocate(20);
state->dataHVACGlobal->TimeStepSys = state->dataGlobal->TimeStepZone;
SetupTimePointers(*state, "Zone", state->dataGlobal->TimeStepZone);
SetupTimePointers(*state, "HVAC", state->dataHVACGlobal->TimeStepSys);

SizingManager::ManageSizing(*state); <-- call from unit test that wound up in UpdateDataandReport

Myoldmopar commented:
@rraustad thanks, I added some of that stuff, but it's still failing.  Somehow the `op->VarMeterArrays(rVar.MeterArrayPtr).NumOnMeters` value is non-zero, but the `state.dataOutputProcessor->MeterValue` array is not being allocated.  So in `UpdateDataandReport`, it's getting an array access violation.  I tried adding a call to `UpdateMeterReporting`, which is where the MeterValue array gets allocated normally, but it didn't do anything.  I can't quite figure out why the code is confused about the number of meters/variables.

Myoldmopar commented:
It looks like both failing unit tests are now hitting the same array out of bounds inside UpdateDataandReport.  I'm not sure how to proceed on this one right now.

rraustad commented:
I'll take a look to see why this still fails with the fix I used before.

Myoldmopar commented:
Thanks @rraustad.  FWIW, there were two unit test failures.  One was an array bounds issue and one was the time pointer map issue.  When I added your code to the time pointer failure, it then become another array bounds issue just like the first.  So locally I get two failures with array bounds in the UpdateDataandReport function.  Just something odd with variables and meters with these changes.

rraustad committed:
Fix unit tests

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 9191-pure-loads-calc

rraustad commented:
@Myoldmopar unit tests run clean locally.

rraustad commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -5875,7 +5879,7 @@ void UpdateDataandReport(EnergyPlusData &state, OutputProcessor::TimeStepType co
                 auto &rVar(op->RVariableTypes(Loop).VarPtr);
                 // Update meters on the TimeStep  (Zone)
-                if (rVar.MeterArrayPtr != 0) {
+                if (rVar.MeterArrayPtr != 0 && !state.dataOutputProcessor->MeterValue.empty()) {
```
I think we do need a new example file to be sure meter data reports correctly when only load calculations are performed.

Myoldmopar replied:
This feels a bit like a band-aid to me.  I am suspicious that we are just short-circuiting some piece of code that is necessary to setup MeterValue, and we should make sure to call that instead of adding this passive check here.  Maybe this is OK as-is, but I'll need to understand it a bit more.

rraustad replied:
I agree with that.

EnergyArchmage replied:
UpdateDataAndReport() is a big beast and it triggers tons of stuff to happen.  It is reused here to get output variables to generate during the ideal loads sizing calcs, but since it now can start getting called much earlier there are a lot of side effects that are not wanted until later, hence the band aids.    

Could make a smaller version of UpdateDataAndReport() to call during ideal loads sizing but that would duplicate code...

rraustad commented:
@EnergyArchmage If I change the SimulationControl inputs in Minimal.idf I get the following error. Is that expected?

    SimulationControl,
      No,                      !- Do Zone Sizing Calculation
      No,                      !- Do System Sizing Calculation
      No,                      !- Do Plant Sizing Calculation
      No,                      !- Run Simulation for Sizing Periods
      No,                      !- Run Simulation for Weather File Run Periods
      No,                      !- Do HVAC Sizing Simulation for Sizing Periods
      1;                       !- Maximum Number of HVAC Sizing Simulation Passes

    Program Version,EnergyPlus, Version 22.2.0-fb84b70ef4, YMD=2022.07.20 18:40,
       ** Warning ** "Do the design day simulations" and "Do the weather file simulation" are both set to "No".  No simulations will be performed, and most input will not be read.
       ** Severe  ** All elements of SimulationControl are set to "No". No simulations can be done.  Program terminates.
       **  Fatal  ** Program terminates due to preceding conditions.
       ...Summary of Errors that led to program termination:
       ..... Reference severe error count=1
       ..... Last severe error=All elements of SimulationControl are set to "No". No simulations can be done.  Program terminates.

rraustad commented:
OK, I see. You have to do at least a zone or system sizing run.

    if ((!state.dataGlobal->DoDesDaySim && !state.dataGlobal->DoWeathSim && !state.dataGlobal->DoHVACSizingSimulation) &&
        (state.dataGlobal->DoZoneSizing || state.dataGlobal->DoSystemSizing)) {
        state.dataGlobal->DoPureLoadCalc = true;
    }

rraustad commented:
I took 5ZoneAirCooled and changed to:

    SimulationControl,
      Yes,                     !- Do Zone Sizing Calculation
      Yes,                     !- Do System Sizing Calculation
      Yes,                     !- Do Plant Sizing Calculation
      No,                      !- Run Simulation for Sizing Periods
      No,                      !- Run Simulation for Weather File Run Periods
      No,                      !- Do HVAC Sizing Simulation for Sizing Periods
      1;                       !- Maximum Number of HVAC Sizing Simulation Passes

and I get a crash here where kEndUseSub == 2:

            for (int jEndUse = 1; jEndUse <= DataGlobalConstantsData::iEndUseSize; ++jEndUse) {
                for (int kEndUseSub = 1; kEndUseSub <= state.dataOutputProcessor->EndUseCategory(jEndUse).NumSubcategories; ++kEndUseSub) {
                    collapsedEndUseSub(kEndUseSub, jEndUse, 1) = ort->gatherEndUseSubBEPS(kEndUseSub, jEndUse, 1);   // electricity

and collapsedEndUseSub is allocated to 1 there:

![image](https://user-images.githubusercontent.com/5778156/180100606-c4a44580-f5bb-478f-a969-76280487735a.png)

rraustad commented:
Looking at [this](https://github.com/NREL/EnergyPlus/pull/9402#issuecomment-1190859460) comment and [this SimulationControl setup](https://github.com/NREL/EnergyPlus/pull/9402#issuecomment-1190871856) shouldn't plant be included in the former test?

rraustad commented:
Documenting this before I move on. It's ort->gatherEndUseSubBEPS that is not allocated correctly.

![image](https://user-images.githubusercontent.com/5778156/180104513-d05a3691-a9be-4c20-b6e4-d961d2732aba.png)

rraustad commented:
Tracked down the allocation.

![image](https://user-images.githubusercontent.com/5778156/180105128-0b02c24b-b63e-4ee4-81e8-4b7f80d0c90d.png)

And more interesting is if I increase the MaxNumSubcategories to 3, this simulation runs. So where is this allocation value determined?

![image](https://user-images.githubusercontent.com/5778156/180107549-ea334869-f98f-425e-816e-1160972bea04.png)

rraustad commented on src/EnergyPlus/OutputProcessor.cc:
```diff
@@ -2467,6 +2467,10 @@ namespace OutputProcessor {
 
+        if (!op->MeterValue.allocated()) {
+            return;
+        }
```
@Myoldmopar note that this new code protects this call before meters are allocated. Similar band-aid since each of the loops below are from 1 to op->NumEnergyMeters. I guess it's possible to know the number of meters before those meters are actually set up. Maybe the new calls (those that change the calling order) could be delayed until something like state.dataGlobal->MetersHaveBeenInitialized = true ? If that works then could unload those meter functions (i.e., remove all new protections) that are called throughout the simulation.

Myoldmopar commented:
@rraustad thanks for looking into this.  FWIW, I do feel like it is really close.  So in a standard E+ run, what function is executed to populate the MeterValue array?  It seems like that piece of setup is missing in our unit test.  Sorry, I'm just a bit reluctant to add more passive checks in the code when hopefully we can just make the right setup call and not have to do these checks.  

Or should the unit test just allocate MeterValue to the right size in order to test this functionality in an isolated fashion?

jmarrec commented:
so I have made a kind of brute force test, testing 4 E+ clis:

* `ORI`: 22.1.0
* `PURE_ONLY`: @EnergyArchmage’s branch `9191-pure-loads-calc ` ( this PR: #9402)
* `NEW`: My branch `9157_9191_SimulationControl_Adjustments` (PR #9386)
* `NEW_PURE`: Merge of both branches + resolve conflicts (`9157_9191_SimulationControl_Adjustments_2`: https://github.com/NREL/EnergyPlus/tree/9157_9191_SimulationControl_Adjustments2)

For each CLI, I'm testing all combinations of Yes/No values of the SimulationControl  object with 5ZoneAirCooled.idf, and looking at the return code from E+: 0 means ok, 1 means it gracefully exited, and anything else indicates a crash.

I was getting some crashes too. I'm rebuilding all three branches, and will rerun the brute force tests and post results here. I'm uploading the notebook I used for the analysis to the EnergyPlusDevSupport repo too.

----

**Edit**: results here: https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9191/results.csv

And the notebook https://github.com/NREL/EnergyPlusDevSupport/blob/master/DefectFiles/9000s/9191/Test_All_Combinations_Simulation_Control.ipynb

This branch is still crashing in release mode with:

```
SimulationControl,
    No,                     !- Do Zone Sizing Calculation
    No,                     !- Do System Sizing Calculation
    Yes,                     !- Do Plant Sizing Calculation
    No,                      !- Run Simulation for Sizing Periods
    No,                      !- Run Simulation for Weather File Run Periods
    No,                      !- Do HVAC Sizing Simulation for Sizing Periods
    1;                       !- Maximum Number of HVAC Sizing Simulation Passes
```

rraustad commented:
@EnergyArchmage 

This new line of code that was added to prevent a crash:

![image](https://user-images.githubusercontent.com/5778156/181818996-6c1402fb-6a96-4826-a741-a76d61ce15c8.png)

was needed because MeterValue is not allocated until UpdateMeterReporting is called and Meters are read in getInput style. This happens after sizing is complete, so the question is do you want the meter data available during load calculations?

![image](https://user-images.githubusercontent.com/5778156/181819407-f5e07f91-a9fd-4412-9dbb-0ed4eca8433c.png)

rraustad commented:
If we only had a function that could read in object data early in the simulation, like the one in #9336 InputProcessor::isInputObjectUsed

EnergyArchmage commented:
There's no need for meters during ideal loads sizing. Many output variables won't be setup yet, so pure load calc users won't be able to get, or care, about many types of data. Reporting goals include, zone heat balance details during ideal loads calcs, component loads, equipment size outcomes

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Richard Raustad ***@***.***>
Sent: Friday, July 29, 2022 2:16:24 PM
To: NREL/EnergyPlus ***@***.***>
Cc: brent.griffith energyarchmage.com ***@***.***>; Mention ***@***.***>
Subject: Re: [NREL/EnergyPlus] Allow running just load calculations and no full simulations (PR #9402)


If we only had a function that could read in object data early in the simulation, like the one in #9336<https://github.com/NREL/EnergyPlus/pull/9336> InputProcessor::isInputObjectUsed

—
Reply to this email directly, view it on GitHub<https://github.com/NREL/EnergyPlus/pull/9402#issuecomment-1199823061>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACCZMUNZ4NUENRVNJIQ5N5LVWQNXRANCNFSM5UPRJBLA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Myoldmopar commented:
I will concede here, let's merge this one in.  I'll pull develop and test/build locally to make sure it's happy with the other PRs that have merged.

Myoldmopar commented:
Everything happy here, thanks for the contribution, and let's move forward with this one.  Merging.
